### DISCUSSION TITLE
Suggestions for bedrock-protocol v4

### DISCUSSION THREAD

**Discussion by hvlxh** [PrismarineJS/bedrock-protocol#378] [Ideas]
1. Use camel/pascel case in `<Client>.write('levelChunk', { hereItIs: "yah" })`
> Use can parse camel case into underscore case by using an function/package

2. Better Types system.
> All types of an file in a separate file
> Enums for `<Client>.on(name)` like names are join, leave, login, etc

3. Working examples
> They are some problems in the examples folder, So it should be fixed.

4. Write whole code in typescript?

**Comment by extremeheat** [comment]
1. Bedrock protocol is a library for other libraries. For example mineflayer will wrap all the network logic and provide an stable API, so we don't worry too much about breaking network changes. These are to be expected. 

2 and 3 do not need a breaking change. Feel free to contribute something. It's better than making issues for trivial things where you know the problem. 

4. See prior discussion in #contribute on Discord. PrismarineJS won't be moving to typescript.

**Comment by extremeheat** [comment]
Changes to the packet names aren't handled in bedrock-protocol, they're handled upstream in minecraft-data. So breaking changes can happen there if corrections need to be made or a new versions requires them, and we'd update usage in node-minecraft-protocol/bedrock-protocol, mineflayer and all the other prismarine- projects accordingly. 

A breaking change in bedrock-protocol would be changes to the API here.

**Comment by ghost** [comment]
I don't really like the 1 suggestion cause it would break a lot of bedrock-protocol and node-minecraft-protocol projects and my server software

**Comment by extremeheat** [comment]
Yeah, and it seems unnecessary to make big changes to the protocol data for that. 

The implication to a bedrock-protocol 4.0 is that you can just use bedrock-protocol@3 to revert the packet changes. That's not the case. I'll move this to a discussion as it's not a present issue in the software to keep the board clean.
